src/luarocks/build/builtin.lua

```diff
@@ -151,27 +143,68 @@ function builtin.run(rockspec)
          return ok, wrapname
       end
    else
-      compile_object = function(object, source, defines, incdirs)
+      compile.dynamic_object = function(object, source, defines, incdirs)
          local extras = {}
-         add_flags(extras, "-D%s", defines)
-         add_flags(extras, "-I%s", incdirs)
+         add_flags(extras, "-D%s", defines, variables)
+         add_flags(extras, "-I%s", incdirs, variables)
          return execute(variables.CC.." "..variables.CFLAGS, "-I"..variables.LUA_INCDIR, "-c", source, "-o", object, unpack(extras))
       end
-      compile_library = function (library, objects, libraries, libdirs)
+      compile.dynamic_library = function (library, objects, libraries, libdirs)
          local extras = { unpack(objects) }
-         add_flags(extras, "-L%s", libdirs)
+         add_flags(extras, "-L%s", libdirs, variables)
          if cfg.gcc_rpath then
-            add_flags(extras, "-Wl,-rpath,%s:", libdirs)
+            add_flags(extras, "-Wl,-rpath,%s:", libdirs, variables)
          end
-         add_flags(extras, "-l%s", libraries)
+         add_flags(extras, "-l%s", libraries, variables)
          if cfg.link_lua_explicitly then
-            add_flags(extras, "-l%s", {"lua"})
+            add_flags(extras, "-l%s", {"lua"}, variables)
          end
          return execute(variables.LD.." "..variables.LIBFLAG, "-o", library, "-L"..variables.LUA_LIBDIR, unpack(extras))
       end
-      compile_wrapper_binary = function(_, name) return true, name end
+      compile.static_object = function(object, source, defines, incdirs)
+         local extras = {}
+         add_flags(extras, "-D%s", defines, variables)
+         add_flags(extras, "-I%s", incdirs, variables)
+         return execute(variables.CC.." "..variables.CFLAGS_STATIC, "-I"..variables.LUA_INCDIR, "-c", source, "-o", object, unpack(extras))
+      end
+      compile.wrapper_binary = function(_, name) return true, name end
       --TODO EXEWRAPPER
    end
+   return compile
+end
+
+local function build_objects(info, compile_object)
+   local objects = {}
+   local sources = info.sources
+   if info[1] then sources = info end
+   if type(sources) == "string" then sources = {sources} end
+   for _, source in ipairs(sources) do
+      local object = source:gsub("%.[^.]*$", "."..cfg.obj_extension)
+      if not object then
+         object = source.."."..cfg.obj_extension
+      end
+      local ok = compile_object(object, source, info.defines, info.incdirs)
+      if not ok then
+         return nil, "Failed compiling object "..object
```

Comment:
So, `build_objects` can return `nil` and error message, but callers do not check it, causing an error later when `nil` is used instead of `objects` array, like in the travis failure.